analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs

```diff
@@ -71,14 +76,21 @@ public string CredentialWords
 
         protected DoNotHardcodeCredentialsBase(IAnalyzerConfiguration configuration)
         {
+            const string Rfc3986_Unreserved = "-._~";  // Numbers and letters are embedded in regex itself without escaping
+            const string Rfc3986_Pct = "%";
+            const string Rfc3986_SubDelims = "!$&'()*+,;=";
+            const string UriPasswordSpecialCharacters = Rfc3986_Unreserved + Rfc3986_Pct + Rfc3986_SubDelims;
+            // See https://tools.ietf.org/html/rfc3986 Userinfo can contain groups: unreserved | pct-encoded | sub-delims
+            var uriUserInfoPart = @"[\w\d" + Regex.Escape(UriPasswordSpecialCharacters) + "]+";
+            uriUserInfoPattern = new Regex(@"\w+:\/\/(?<Login>" + uriUserInfoPart + "):(?<Password>" + uriUserInfoPart + ")@", RegexOptions.Compiled);
```

Comment:
I don't understand the comment. Can you provide more detailed explanation?

Reply:
I did not really changed anything of this logic, I just moved it to different place

Reply:
I think this is a valid userinfo that does not match the regex:
`https://user:passwordWith:isOkay@example.org/` This is how I read the grammar in the RFC, but I could be mistaken. At least .Net thinks so too [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBfsQJwAUFABM4zAKYDujAqslHQCIAFggQAHAM4gA9NNQS2yEGIgSJHAPbIMAdSgIhIKBIDyAawgBPfPjYAPCGDEAbNgDotAc2kCAlG+4FZABJGAAzDQpfAG4gA==)

Reply:
We don't need to extract the full password. So changing it will make the regex more expensive, without affecting the results.

Reply:
The problem is that the `:` is not in the `<Password>` group and because of the trailing `@`, the regex doesn't match at all. The password group must be `"[\w\d" + Regex.Escape(UriPasswordSpecialCharacters) + ":]+"`.

Reply:
I'll try in another PR

Reply:
See #6199